### DISCUSSION TITLE
Массив недоступных треков и отслеживание их доступности

### DISCUSSION THREAD

**Discussion by Bondian** [Chimildic/goofy#198] [Помощь с алгоритмом] [answered]
Хочу создать файл на гугл-диске, куда бы сливались все недоступные треки, встречающиеся при фильтрации в разных функциях. По еженедельному триггеру проверять этот файл на наличие треков, ставших доступными, и помещать эти треки в отдельный плейлист.
Для начала вопрос. Есть ли более простой способ выделить недоступные треки в массив, коме как таким способом:
```
  let tracks = Source.getPlaylistTracks('', 'id');
  let tracksCopy = Selector.sliceCopy(tracks);
  Filter.removeUnavailable(tracksCopy);
  let unavailableTracks = Filter.removeTracks(tracks, tracksCopy);
```

**Comment by Chimildic** [comment, marked as answer]
Всю логику сохранения можно вынести в отдельную функцию и вызывать ее при сборе любого плейлиста одной строкой.
```js
  let followedTracks = Source.getFollowedTracks({ type: 'owned' });
  // ...
  appendUnavailableTracks_(followedTracks)
```
Побочным эффектом этой функции будет удаление недоступных треков из оригинального массива. То есть фильтр `removeUnavailable` на вызывающей стороне можно удалить.
```js
function appendUnavailableTracks_(source) {
  let unavailable = Selector.sliceCopy(source)
  Filter.removeUnavailable(source)
  Filter.removeTracks(unavailable, source)
  if (unavailable.length > 0) {
    Cache.compressTracks(unavailable)
    Cache.append("unavailable.json", unavailable)
  }
}
```
Возможно имеет смысл изменить `appendUnavailableTracks_` так, чтобы кэш записывался после удаления дубликатов.

**Reply by Bondian** [reply]
> Побочным эффектом этой функции будет удаление недоступных треков из оригинального массива.

Я правильно понимаю, что если применять фильтр `removeUnavailable` к копии, то оригинальный массив не будет затронут?
```
  function appendUnavailableTracks_(source) {
    let unavailable = Selector.sliceCopy(source)
    let sourceCopy = Selector.sliceCopy(source)
    Filter.removeUnavailable(sourceCopy)
    Filter.removeTracks(unavailable, sourceCopy)
    if (unavailable.length > 0) {
      Cache.compressTracks(unavailable)
      Cache.append("unavailable.json", unavailable)
    }
```
(не знаю как здесь подсветить код разным цветом, чтобы он выглядел также достойно)
Ещё не понял кое-что... В создаваемой функции точку с запятой не нужно ставить после каждой строки?

**Reply by Chimildic** [reply]
Да, в случае копии оригинал не изменяется. Но здесь побочный эффект на руку, как по мне.
Точка с запятой необязательна для JavaScript. Есть некоторая экзотика, которая может привести к ошибке. Но сколько писал на goofy никогда не попадалось подобных случаев. Поэтому это больше вопрос к стилю. Либо везде, либо нигде.

**Reply by Bondian** [reply]
Спасибо! Буду внедрять.
